org-ql.el

```diff
@@ -448,6 +454,41 @@ Returns cons (INHERITED-TAGS . LOCAL-TAGS)."
                  org-ql-tags-cache))
       (puthash position all-tags tags-cache))))
 
+(defun org-ql--expand-tag-hierarchy (tags &optional excluded)
+  "Return TAGS along with their associated group tags.
+This function recursively searches for groups that each given tag belongs to,
+directly or indirectly, and includes the corresponding group tags to the result.
+
+TAGS should be a list of tags (i.e., strings).
+If non-nil, EXCLUDED should be a list of group tags that will not be
+automatically added to the results unless they are already in TAGS."
+  (let ((groups (org-tag-alist-to-groups org-current-tag-alist))
+        (excluded (append tags excluded))
+        result)
+    (let (group-tags)
+      (dolist (tag tags)
+        (pcase-dolist (`(,group-tag . ,group-members) groups)
+          (when (and (not (member group-tag excluded))
+                     ;; Check if one of the members in the group matches tag.
+                     ;; Notice that each member may be a plain string or
+                     ;; a regexp pattern (enclosed between curly brackets).
+                     (--some (if (string-match-p "^[{].+[}]$" it)
+                                 ;; If pattern (it) is a regexp, remove the brackets and
+                                 ;; make sure that it either matches the whole tag or not.
+                                 (string-match-p (concat "^" (substring it 1 -1) "$") tag)
```

Comment:
Again, I haven't used group tags, so I'll mostly defer to your knowledge here.  But I'm curious, is it necessary to wrap the regexp with `^` and `$` instead of testing it as-is?  Is Org explicit about what is intended here?

Reply:
I haven't looked into the actual implementation of this in org but judging from how `org-sparse-tree` behaves during tag/property matching, then the whole tag must match the regexp.

Reply:
Okay, then.  I wish we could avoid concatting new strings, but if we have to, we have to.  Thanks.